File: boj_2504/solution.cpp
#include <bits/stdc++.h>
using namespace std;

char oppositBracket(char closing) {
  switch (closing) {
   case ')': return '(';
   case ']': return '[';
   default: throw;
  }
}

int value(char closing) {
  switch (closing) {
   case ')': return 2;
   case ']': return 3;
   default: throw;
  }
}

void solve() {
  string input;
  cin >> input;
  
  stack<char> S;
  for (char c : input) {
    if (c == '(' || c == '[') {
      S.push(c);
      continue;
    }
    int x = 0;
    char o = oppositBracket(c);
    while (S.top() != o) {
      cout << S.top() << ", " << o << ", " << c << "\n";
      if (!isdigit(S.top())) {
        printf("0\n");
        return;
      }
      x += S.top();
      S.pop();
    }
    S.pop();
    S.push(value(c) * (x == 0 ? 1 : x));
  }
  
  int ans = 0;
  while (!S.empty()) {
    ans += S.top();
    S.pop();
  }
  printf("%d\n", ans);
}

int main() {
  int T = 1;
  while (T--) {
    solve();
  }

  return 0;
}

File: boj_2504/solution.py
import sys
input = sys.stdin.readline

def opposite(closing):
    return {
        ')': '(',
        ']': '[',
    }[closing]

def value(closing):
    return {
        ')': 2,
        ']': 3,
    }[closing]

def is_opening(c):
    return c in '(['

def is_bracket(c):
    return isinstance(c, str) and c in '([])'

def solve():
    stack = []
    for c in input().strip():
        if not is_bracket(c):
            return 0
        if is_opening(c):
            stack.append(c)
            continue
        o = opposite(c)
        if not stack:
            return 0
        x = 0
        while stack and stack[-1] != o:
            if not isinstance(stack[-1], int):
                return 0
            x += stack.pop()
        if not stack or stack[-1] != o:
            return 0
        stack.pop()
        stack.append(value(c) * (1 if x == 0 else x))
    if any(is_bracket(c) for c in stack):
        return 0
    return sum(stack)

print(solve())
